api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java

```diff
@@ -27,908 +28,1029 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public NDArray toDevice(Device device, boolean copy) {
+        if (device.equals(getDevice()) && !copy) {
+            return this;
+        }
+        return duplicate();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public NDArray toType(DataType dataType, boolean copy) {
+        if (dataType.equals(getDataType()) && !copy) {
+            return this;
+        }
+        return duplicate();
     }
 
     /** {@inheritDoc} */
     @Override
-    default void setRequiresGradient(boolean requiresGrad) {
+    public void setRequiresGradient(boolean requiresGrad) {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray getGradient() {
+    public NDArray getGradient() {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default boolean hasGradient() {
+    public boolean hasGradient() {
         return false;
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray stopGradient() {
+    public NDArray stopGradient() {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default String[] toStringArray() {
+    public String[] toStringArray() {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default void set(Buffer data) {
+    public void set(Buffer data) {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default void copyTo(NDArray array) {
+    public void copyTo(NDArray array) {
         throw new UnsupportedOperationException(UNSUPPORTED_MSG);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray booleanMask(NDArray index, int axis) {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public NDArray booleanMask(NDArray index, int axis) {
+        return getAlternativeArray().booleanMask(index, axis);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray sequenceMask(NDArray sequenceLength, float value) {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public NDArray sequenceMask(NDArray sequenceLength, float value) {
+        return getAlternativeArray().sequenceMask(sequenceLength, value);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray sequenceMask(NDArray sequenceLength) {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public NDArray sequenceMask(NDArray sequenceLength) {
+        return sequenceMask(sequenceLength, 0);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray zerosLike() {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public NDArray zerosLike() {
+        return getManager().zeros(getShape(), getDataType(), getDevice());
```

Comment:
Some of these are fully generic and don't rely on anything adapter specific. We may want to move the definitions to NDArray instead. This could be in another PR if you prefer.

Reply:
I didn't move to NDArray is because I'm not sure if we should remove native implementation for MXNet and PyTorch

Reply:
We don't need to remove the native implementations. The ones in NDArray are just defaults and can be overridden for better performance

Reply:
I don't think there is performance gain, the only thing it might saved is getShape() and. getDataType() call.